torch/csrc/jit/pickle.cpp

```diff
@@ -38,58 +36,74 @@ std::vector<char> pickle(
 // This has to live here instead of the C++ API to mirror torch.save since the
 // mobile build excludes the C++ API
 std::vector<char> pickle_save(const at::IValue& ivalue) {
-  std::vector<char> data;
-
-  auto writer = [&](const char* bytes, size_t len) {
-    data.insert(data.end(), bytes, bytes + len);
-  };
-
-  jit::Pickler pickler(writer, /*tensor_table=*/nullptr);
-  // Output data to match torch.save, see torch/serialization.py for details
-  // Magic number (0x1950a86a20f9469cfc6c)
-  pickler.protocol();
-  pickler.pushLong(torch_save_magic_number);
-  pickler.stop();
-
-  // Protocol Version
-  pickler.protocol();
-  pickler.pushInt(protocol_version);
-  pickler.stop();
-
-  // sys_info, this isn't actually used in de-serialization so we can leave this
-  // one empty
+  // Pickle the IValue into an array of bytes
+  std::vector<char> pickle_data;
+  Pickler pickler(
+      [&](const char* buf, size_t size) {
+        pickle_data.insert(pickle_data.end(), buf, buf + size);
+      },
+      /*tensor_table=*/nullptr,
+      /*class_table=*/nullptr);
   pickler.protocol();
-  pickler.pushEmptyDict();
+  pickler.pushIValue(ivalue);
   pickler.stop();
 
-  jit::Pickler data_pickler(writer, /*tensor_table=*/nullptr);
-  data_pickler.protocol();
-  data_pickler.pushIValue(ivalue);
-  data_pickler.stop();
-
-  auto writeable_tensors = data_pickler.tensorData();
-
-  std::vector<at::IValue> keys;
-  keys.reserve(writeable_tensors.size());
-  std::vector<at::TypePtr> types(writeable_tensors.size(), at::StringType::get());
-
-  for (size_t i = 0; i < writeable_tensors.size(); i++) {
-    keys.emplace_back(c10::to_string(i));
-  }
+  std::vector<char> container_data;
+  container_data.reserve(pickle_data.size() );
 
-  auto keys_tuple = at::ivalue::Tuple::create(keys);
-  jit::pickle(writer, keys_tuple);
-
-  for (const auto& tensor_data : writeable_tensors) {
-    const char* addr = tensor_data.data();
-    size_t numel = tensor_data.numel();
-    writer(reinterpret_cast<const char*>(&numel), sizeof(numel));
-    writer(addr, tensor_data.sizeInBytes());
-  }
-
-  return data;
+  caffe2::serialize::PyTorchStreamWriter writer(
+      [&](const void* void_bytes, size_t len) {
+        const char* bytes = reinterpret_cast<const char*>(void_bytes);
+        container_data.insert(container_data.end(), bytes, bytes + len);
+        return len;
+      });
+
+  // Write the generated bytes and the associated tensors into a data.pkl file
+  // and data/0, data/1, data/2... files for each of the tensors
+  writeArchiveAndTensors(
```

Comment:
This breaks mobile CI because export.cpp is not included in mobile build:
https://github.com/pytorch/pytorch/blob/master/caffe2/CMakeLists.txt#L488

One simple work around is to gate this logic with "#ifndef C10_MOBILE .... #else AT_ERROR("not supported on mobile"); #endif"
Ideally we probably should move saving functionality to a separate file so that we can optionally include/exclude this function at link time. cc: @dreiss who might need mobile saving function for federate learning.
But I think you can go with the simple work around to unblock yourself for now.